yellow-paper/docs/contract-deployment/instances.md

```diff
@@ -17,102 +17,185 @@ The structure of a contract instance is defined as:
+
+### Allowing Uninitialized Publicly Deployed Instances
+
+The `ContractInstanceDeployer` requires that a contract is Initialized before agreeing to Publicly Deploy it, either by checking that the Initialization Nullifier has been already emitted, or by invoking a constructor itself and checking immediately after. 
+
+Removing this check greatly simplifies the `ContractInstanceDeployer`, since it does not need to invoke constructors or check initialization nullifiers emitted from other contracts.
+
+```diff
+function deploy (
+  user_salt: Field, 
+  contract_class_id: Field, 
+  initialization_hash: Field, 
+  portal_contract_address: Field, 
+  public_keys_hash: Field, 
+  universal_deploy?: boolean,
+-  constructor_selector?: Field,
+-  constructor_args?: Field[],
+)
+  assert nullifier_exists silo(contract_class_id, ClassRegisterer)
+  assert is_valid_eth_address(portal_contract_address)
+  
+  deployer = if universal_deploy then this.address else msg_sender
+  salt = pedersen([user_salt, this.address], GENERATOR__ADDRESS_SALT)
+  version = 1
+  address = compute_address(version, salt, contract_class_id, initialization_hash, portal_contract_address, public_keys_hash)
+
+  emit_nullifier(address)
+-  if constructor_selector then call(address, constructor_selector, constructor_args)
+-  assert nullifier_exists silo(address, address)
+
+  emit_event ContractInstanceDeployed(address, version, salt, contract_class_id, initialization_hash, portal_contract_address, public_keys_hash)
+```
+
+This approach also allows devs to handle initialization in alternate ways, since there is nothing in the protocol or in a canonical contract that mandates how contracts are expected to be initialized. And it also allows for contracts to _not_ have a constructor at all if it is not needed, while still allowing them to be Publicly Deployed.
+
+However, this opens the door for a contract to be simultaneously Publicly Deployed and Uninitialized, which is a state that does not seem to map to a valid use case. Furthermore, it requires public functions to check the Initialization Nullifier on every call, which in the current approach is not needed as the presence of the Deployment Nullifier checked by the Public Kernel is enough of a guarantee that the contract was initialized.
+
+<!-- To @iAmMichaelConnor: I am tempted to actually implement this approach in spite of its drawbacks. It makes responsibilities so much more clearly defined: Deployer deploys, Initializer initializes, and there is no link between the two. It also allows removing constructors when they are not needed. And the added cost of checking the initialization nullifier from public functions doesn't seem too terrible. -->
```

Comment:
> Is the purpose of "deploy" solely to enforce the bytecode to be broadcast?

Yep!

>  If the deployment nullifier does exist, the sequencer doesn't need to prove that fact, because they clearly have the bytecode and they can prove this bytecode is correct via the initialisation nullifier.

Not sure I agree. Even if the sequencer can prove that they know the bytecode, it's important they prove that _everyone_ knows the bytecode of the public function being executed. Otherwise, we open the door to public functions that can only be run by a subset of nodes that are privy to that bytecode, which doesn't feel right.

Reply:
> If the merkle membership/non-membership check for the deployment nullifier is done in the kernel circuit, then this observation doesn't help much: the kernel circuit is a fixed circuit with no branching, so the hashing will always happen.

Good point! Given this, I'm even more tempted to go down this route and completely split initialization and deployment.

> But if there's a way for the hashing to be done within the AVM, then the sequencer can skip the merkle membership proof unless they want to claim gas fees for a non-deployed contract.

Not sure. Feels doable, but it also feels like we're pushing into the AVM a responsibility that is very clearly the kernel's.